elephant/spike_train_generation.py

```diff
@@ -18,6 +18,104 @@
+    time_stamps: None, quantity array or Object with .times interface
+        if 'spike_train' is a quantity array or exposes a quantity array
+        exposes the .times interface, it provides the time_stamps
+        around which the waveform is extracted. If it is None, the
+        function peak_detection is used to calculate the time_stamps
+        from signal. Default: None.
+    extr_interval: unpackable time quantities, len == 2
+        'extr_interval' specifies the time interval around the
+        time_stamps where the waveform is extracted. The default is an
+        interval of '6 ms'. Default: (-2 * ms, 4 * ms).
+
+    Returns
+    -------
+    result_st : neo SpikeTrain object
+        'result_st' contains the time_stamps of each of the spikes and
+        the waveforms in result_st.waveforms.
+    """
+    # Get spike time_stamps
+    if time_stamps is None:
+        time_stamps = peak_detection(signal, threshold, sign=sign)
+    elif hasattr(time_stamps, 'times'):
+        time_stamps = time_stamps.times
+    elif type(time_stamps) is Quantity:
+        raise TypeError("time_stamps must be None, a quantity array or" +
+                        " expose the.times interface")
+
+    if len(time_stamps) == 0:
+        return SpikeTrain(time_stamps, units=signal.times.units,
+                          t_start=signal.t_start, t_stop=signal.t_stop,
+                          waveforms=np.array([]),
+                          sampling_rate=signal.sampling_rate)
+
+    # Unpack the extraction interval from tuple or array
+    extr_left, extr_right = extr_interval
+    if extr_left > extr_right:
+        raise ValueError("extr_interval[0] must be < extr_interval[1]")
+
+    if any(np.diff(time_stamps) < extr_interval[1]):
+        print("WARNING: Waveforms overlap")
+
+    data_left = ((extr_left * signal.sampling_rate).simplified).magnitude
+
+    data_right = ((extr_right * signal.sampling_rate).simplified).magnitude
+
+    data_stamps = (((time_stamps - signal.t_start) *
+                    signal.sampling_rate).simplified).magnitude
+
+    data_stamps = data_stamps.astype(int)
+
+    borders_left = data_stamps + data_left
+
+    borders_right = data_stamps + data_right
+
+    borders = np.dstack((borders_left, borders_right)).flatten()
+
+    waveforms = np.array(
+                np.split(np.array(signal), borders)[1::2]) * signal.units
+
+    # len(np.shape(waveforms)) == 1 if waveforms do not have the same width.
+    # this can occur when extr_interval indexes beyond the signal.
+    # Workaround: delete spikes shorter than the maximum length with
+    if len(np.shape(waveforms)) == 1:
+        max_len = (np.array([len(x) for x in waveforms])).max()
+        to_delete = np.array([idx for idx, x in enumerate(waveforms)
+                             if len(x) < max_len])
+        waveforms = np.delete(waveforms, to_delete, axis=0)
+        waveforms = np.array([x for x in waveforms])
+        print("WARNING: Waveforms " +
```

Comment:
I did not use the warning module because it was not imported in spike_train_generation.py when I started working and I did not want to introduce additional imports. But I can use the warning module if it is ok to import.